File: Android/NatReader/src/main/java/api/natsuite/natreader/MP4Reader.java
package api.natsuite.natreader;

import android.media.MediaExtractor;
import android.media.MediaFormat;
import android.util.Log;
import java.io.IOException;

public final class MP4Reader implements FrameReader {

    //region --Client API--

    public MP4Reader (String uri) {
        this.uri = uri;
        this.extractor = new MediaExtractor();
        MediaFormat format = null;
        try {
            extractor.setDataSource(uri);
            format = videoFormat(extractor);
            if (format == null)
                Log.e("NatSuite", "NatReader Error: MP4FrameReader failed to find video track in media file");
        } catch (IOException ex) {
            Log.e("NatSuite", "NatReader Error: MP4FrameReader failed to create media extractor with error: " + ex);
        }
        this.format = format;
        if (this.format != null)
            Log.d("NatSuite", "NatReader: Initialized MP4FrameReader for video with duration "+duration()+"s and size "+frameWidth()+"x"+frameHeight()+"@"+frameRate()+"Hz");
    }

    @Override
    public String uri () {
        return uri;
    }

    @Override
    public float duration () {
        return format != null ? format.getLong(MediaFormat.KEY_DURATION) / 1e+6f : 0;
    }

    @Override
    public int frameWidth () {
        return format != null ? format.getInteger(MediaFormat.KEY_WIDTH) : 0;
    }

    @Override
    public int frameHeight () {
        return format != null ? format.getInteger(MediaFormat.KEY_HEIGHT) : 0;
    }

    @Override
    public float frameRate () {
        return format != null ? format.getInteger(MediaFormat.KEY_FRAME_RATE) : 0;
    }

    @Override
    public MediaEnumerator createEnumerator (float startTime, float duration, int frameSkip) {
        if (format != null)
            return new FrameEnumerator(extractor, format, startTime, duration, frameSkip);
        else
            return null;
    }

    @Override
    public void release () {
        extractor.release();
    }
    //endregion


    //region --Operations--

    private final String uri;
    private final MediaExtractor extractor;
    private final MediaFormat format;

    private static MediaFormat videoFormat (MediaExtractor extractor) {
        // Search
        for (int i = 0; i < extractor.getTrackCount(); i++) {
            final MediaFormat format = extractor.getTrackFormat(i);
            if (format.getString(MediaFormat.KEY_MIME).startsWith("video/")) {
                extractor.selectTrack(i);
                return format;
            }
        }
        // Failed
        return null;
    }
    //endregion
}

File: Runtime/MP4Reader.cs
/* 
*   NatReader
*   Copyright (c) 2021 Yusuf Olokoba.
*/

namespace NatSuite.Readers {

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Runtime.InteropServices;
    using System.Text;
    using UnityEngine;
    using Internal;

    /// <summary>
    /// MP4 video frame reader.
    /// </summary>
    public sealed class MP4Reader : IFrameReader {

        #region --Client API--
        /// <summary>
        /// Media source path.
        /// </summary>
        public string path {
            get {
                var result = new StringBuilder(1024);
                reader.Path(result);
                return result.ToString();
            }
        }

        /// <summary>
        /// Media duration in seconds.
        /// </summary>
        public float duration => reader.Duration();

        /// <summary>
        /// Frame size.
        /// </summary>
        public (int width, int height) frameSize {
            get {
                reader.FrameSize(out var width, out var height);
                return (width, height);
            }
        }

        /// <summary>
        /// Frame rate.
        /// </summary>
        public float frameRate => reader.FrameRate();
        
        /// <summary>
        /// Create an MP4 frame reader.
        /// </summary>
        /// <param name="path">Path to MP4 video file.</param>
        public MP4Reader (string path) => this.reader = Bridge.CreateMP4Reader(path);

        /// <summary>
        /// Read frames in a time range.
        /// </summary>
        /// <param name="startTime">Time to start reading samples in seconds.</param>
        /// <param name="duration">Duration in seconds. Pass `-1` to read till the end of the stream.</param>
        /// <param name="frameSkip">Number of frames to skip when reading.</param>
        public IEnumerable<(byte[] pixelBuffer, long timestamp)> Read (float startTime = 0, float duration = -1, int frameSkip = 0) {
            // Create enumerator
            var enumerator = reader.CreateEnumerator(startTime, duration, frameSkip);
            if (enumerator == IntPtr.Zero) {
                Debug.LogError("NatReader Error: MP4FrameReader failed to create enumerator");
                yield break;
            }
            // Read
            var nativeBuffer = Marshal.AllocHGlobal(frameSize.width * frameSize.height * 4); // Workaround for Android
            var pixelBuffer = new byte[frameSize.width * frameSize.height * 4];
            try {
                for (;;) {
                    enumerator.CopyNextFrame(nativeBuffer, out var _, out var timestamp);
                    Marshal.Copy(nativeBuffer, pixelBuffer, 0, pixelBuffer.Length);
                    if (timestamp < 0L) // EOS
                        break;
                    yield return (pixelBuffer, timestamp);
                }
            }
            // Dispose enumerator
            finally {
                Marshal.FreeHGlobal(nativeBuffer);
                enumerator.DisposeEnumerator();
            }
        }

        /// <summary>
        /// Dispose the reader and release resources.
        /// </summary>
        public void Dispose () => reader.Dispose();
        #endregion


        #region --Operations--

        private readonly IntPtr reader;

        public override string ToString () =>
        $"{{{Environment.NewLine}\t" +
        string.Join(
            $"{Environment.NewLine}\t",
            new Dictionary<string, object> {
                ["path"] = path,
                ["duration"] = duration,
                ["frameSize"] = frameSize,
                ["frameRate"] = frameRate
            }.Select(pair => $"{pair.Key}: {pair.Value}")
        ) +
        $"{Environment.NewLine}}}";
        #endregion
    }
}